File: 153-find-minimum-in-rotated-sorted-array/153-find-minimum-in-rotated-sorted-array.java
class Solution {
    public int findMin(int[] nums) {
        int n= nums.length;
        int l=0;
        int h= n-1;
        
        
        
        while(l<=h)
        {
            int mid =l+ (h-l)/2;
            if(nums[l]<=nums[h])
            return nums[l];
            
            int nextIndex= (mid+1)%n;
            int prevIndex = (mid-1+n)%n;
            
            if(nums[mid]<=nums[nextIndex] && nums[mid]<=nums[prevIndex])
                return nums[mid];
            
            else if(nums[mid]>=nums[l])
                l=mid+1;
            else
                h=mid-1;
        }
        return 0;
    }
}

File: 153-find-minimum-in-rotated-sorted-array/153-find-minimum-in-rotated-sorted-array.cpp
class Solution {
public:
    int findMin(vector<int>& nums) {
//              int lo = 0;
//         int hi = nums.size() - 1;
//         int mid = 0;
        
//         while(lo < hi) {
//             mid = lo + (hi - lo) / 2;
            
//             if (nums[mid] > nums[hi]) {
//                 lo = mid + 1;
//             }
//             else if (nums[mid] < nums[hi]) {
//                 hi = mid;
//             }
//             else { // when num[mid] and num[hi] are same
//                 hi--;
//             }
//         }
//         return nums[lo];
        int l=0;
        int h=nums.size()-1;
        if(nums[l]<=nums[h]) return nums[l];
        while(l<=h)
        {
            int mid = (l+h)/2;
            
//             right check
            if(nums[mid]>nums[mid+1]) return nums[mid+1];
            
//             left check
            else if(nums[mid]<nums[mid-1]) return nums[mid];
            
            else if(nums[l]<=nums[mid]) l=mid+1;
            else if(nums[mid]<=nums[h]) h=mid-1;
        }
        return -1;
        
            
    }
};